toolkit/scripts/toolchain.mk

```diff
@@ -171,59 +171,66 @@ $(final_toolchain): $(raw_toolchain) $(STATUS_FLAGS_DIR)/build_toolchain_srpms.f
+else
+# This may be an empty string, that is fine. If its empty we will try to use online packages
+selected_toolchain_archive = $(TOOLCHAIN_ARCHIVE)
+endif
 
+# Extract RPMs from an archive if available, download them from online otherwise.
+ifneq (,$(selected_toolchain_archive))
 # Our manifest files should always track the contents of the freshly built archives exactly
 #   Currently non-blocking, to make this a blocking check switch to `print_error` instead of
 #   `print_warning`
-$(STATUS_FLAGS_DIR)/toolchain_verify.flag: $(TOOLCHAIN_MANIFEST) $(final_toolchain)
-	@echo Validating contents of toolchain against manifest... && \
-	tar -tzf $(final_toolchain) | grep -oP "[^/]+rpm$$" | sort > $(toolchain_actual_contents) && \
+$(STATUS_FLAGS_DIR)/toolchain_verify.flag: $(TOOLCHAIN_MANIFEST) $(selected_toolchain_archive)
+	@echo Validating contents of toolchain against manifest...
+	tar -I $(ARCHIVE_TOOL) -tf $(selected_toolchain_archive) | grep -oP "[^/]+rpm$$" | sort > $(toolchain_actual_contents) && \
 	sort $(TOOLCHAIN_MANIFEST) > $(toolchain_expected_contents) && \
 	diff="$$( comm -3 $(toolchain_actual_contents) $(toolchain_expected_contents) --check-order )" && \
-	echo Toolchain manifest validation complete. && \
 	if [ -n "$${diff}" ]; then \
-		echo "ERROR: Missmatched packages between '$(TOOLCHAIN_MANIFEST)' and '$(final_toolchain)':" && \
+		echo "ERROR: Mismatched packages between '$(TOOLCHAIN_MANIFEST)' and '$(selected_toolchain_archive)':" && \
 		echo "$${diff}"; \
-		$(call print_warning, $@ failed) ; \
-	fi
+		$(call print_error, $@ failed) ; \
+	fi && \
 	touch $@
+	@echo Done validating toolchain
 
+# Extract a set of RPMS from an archive instead of building them from scratch.
+$(toolchain_local_temp): ;
+$(toolchain_local_temp)%: ;
 # The basic set of RPMs can always be produced by bootstrapping the toolchain.
 # Try to skip extracting individual RPMS if the toolchain step has already placed
 # them into the RPM folder.
-$(toolchain_rpms): $(TOOLCHAIN_MANIFEST) | $(final_toolchain)
-	@echo Extracting RPM $@ from toolchain && \
-	if [ ! -f $@ -o $(final_toolchain) -nt $@ ]; then \
-		mkdir -p $(dir $@) && \
-		tar -I $(ARCHIVE_TOOL) -xvf $(final_toolchain) -C $(dir $@) --strip-components 1 built_rpms_all/$(notdir $@) && \
-		touch $@ ; \
-	fi || $(call print_error, $@ failed) ; \
-	touch $@
-else
-ifneq (,$(TOOLCHAIN_ARCHIVE))
-# Extract a set of RPMS from an archive instead of building them from scratch.
-$(toolchain_local_temp): $(STATUS_FLAGS_DIR)/toolchain_local_temp.flag
-	@touch $@
 
-$(toolchain_local_temp)%: ;
-$(STATUS_FLAGS_DIR)/toolchain_local_temp.flag: $(TOOLCHAIN_ARCHIVE) $(shell find $(toolchain_local_temp)/* 2>/dev/null)
-	mkdir -p $(BUILD_DIR)/toolchain_temp/ && \
-	tar -I $(ARCHIVE_TOOL) -xvf $(TOOLCHAIN_ARCHIVE) -C $(BUILD_DIR)/toolchain_temp/ --strip-components 1 && \
+# The $(depend_TOOLCHAIN_ARCHIVE) and $(depend_REBUILD_TOOLCHAIN) argument trackers are used to force a regeneration of the toolchain
```

Comment:
It may not involve actually building anything, this just goes through and makes sure all the .rpm files we expect are available where they are meant to be. I'll add a better comment to that effect.